torch/csrc/cuda/memory_snapshot.cpp

```diff
@@ -105,17 +105,66 @@ void _record_memory_history(
     int64_t trace_alloc_max_entries,
     bool trace_alloc_record_context,
     bool record_cpp_context) {
-  c10::cuda::CUDACachingAllocator::CreateContextFn recorder = nullptr;
-  if (record_context) {
-    if (record_cpp_context) {
-      recorder = gather_with_cpp;
-    } else {
-      recorder = gather;
+  c10::cuda::CUDACachingAllocator::CreateContextFn recorder = gather;
+  if (record_cpp_context) {
+    recorder = gather_with_cpp;
+    if (enabled) {
+      // warm up C++ stack gathering
+      unwind::unwind();
+    }
+  }
+  auto when = c10::cuda::CUDACachingAllocator::RecordContext::NEVER;
+  if (trace_alloc_record_context) {
+    when = c10::cuda::CUDACachingAllocator::RecordContext::ALLOC;
+  } else if (record_context) {
+    when = c10::cuda::CUDACachingAllocator::RecordContext::STATE;
+  }
+  at::globalContext().lazyInitCUDA();
+  c10::cuda::CUDACachingAllocator::recordHistory(
+      enabled, recorder, trace_alloc_max_entries, when);
+}
+
+void _record_memory_history(
+    c10::optional<std::string> enabled,
+    c10::optional<std::string> context,
+    std::string stacks,
+    size_t max_entries) {
+  // if enabled not in ["state", "all", None]:
+  //       raise TypeError("expected state to be 'state', 'all', or None")
+  //   if context not in ["state", "all", None]:
+  //       raise TypeError("expected context to be 'state', 'all', or None")
+  //   if stacks not in ["python", "all"]:
+  //       raise TypeError("expected stacks to be 'python', or 'all'")
+
+  //   enabled_ = enabled is not None
+  //   record_context = context is not None
+  //   trace_alloc_max_entries = max_entries if enabled == "all" else 1
+  //   trace_alloc_record_context = context == "all"
+  //   record_context_cpp = stacks == "all"
+
+  bool e = bool(enabled);
+  c10::cuda::CUDACachingAllocator::CreateContextFn recorder = gather;
+  if (stacks == "all") {
+    recorder = gather_with_cpp;
+    if (e) {
+      // warm up C++ stack gathering
+      unwind::unwind();
+    }
+  }
+  max_entries = (enabled && *enabled == "all") ? max_entries : 1;
+  auto when = c10::cuda::CUDACachingAllocator::RecordContext::NEVER;
+  if (context) {
+    if (context == "all") {
```

Comment:
nit: you can also just bind the enum if you want to avoid this/reduce string on the API: https://github.com/pytorch/pytorch/blob/ad22f0ffb456fc3f967ad32e09376f7c9cf94a56/torch/csrc/Module.cpp#L1496-L1519

Reply:
I want to avoid including the allocator header file in memory_snapshot.h because it slowdown recompilation of the allocator. I also think it makes sense that the C++ and Python API for recording history has the same types.